crates/sqlbuiltins/src/functions/bson.rs

```diff
@@ -0,0 +1,177 @@
+                    // MongoDB.
+                    .max_frame_length(16 * 1024 * 1024)
+                    .length_field_type::<u32>() // actually signed int32s
+                    .length_field_offset(0) // length field is first
+                    .length_adjustment(4) // length prefix includes
+                    // the prefix use the object_store buffered reader
+                    // to stream data from the object store:
+                    .new_read(object_store::buffered::BufReader::with_capacity(
+                        store.to_owned(),
+                        &obj,
+                        32 * 1024 * 1024, // 32 MB buffer, probably still too small
+                    ))
+                    // convert the chunk of bytes to bson.
+                    .map(
+                        // TODO: this probably wants to be a raw document eventually, so we can put
+                        // all the _fields_ in a map, iterate over the document once, and check each
+                        // bson field name against the schema, and only pull out the fields that
+                        // match. This is easier in the short term but less performant for large
+                        // documents where the docuemnts are a superset of the schema, we'll end up
+                        // doing much more parsing work than is actually needed for the bson
+                        // documents.
+                        |bt: Result<BytesMut, std::io::Error>| -> Result<bson::Document, ExtensionError> {
+                            Ok(bson::de::from_slice::<bson::Document>(
+                                bt?.freeze().as_bytes().to_owned().as_slice(),
+                            )?
+                            .to_owned())
+                        }
+                            as fn(
+                                Result<BytesMut, std::io::Error>,
+                            ) -> Result<bson::Document, ExtensionError>,
+                    ),
+            );
+        }
+
+        // iterate through the readers and build up a sample of the first <n> documents to be used
+        // to infer the schema.
+        let mut sample = Vec::<bson::Document>::with_capacity(sample_size as usize);
+        let mut first_active: usize = 0;
+        'readers: for reader in readers.iter_mut() {
+            while let Some(res) = reader.next().await {
+                match res {
+                    Ok(doc) => sample.push(doc),
+                    Err(e) => return Err(e),
+                };
+
+                if sample.len() as i64 >= sample_size {
+                    break 'readers;
+                }
+            }
+            first_active += 1;
+        }
+
+        // if we had to read through one or more than of the input files in the glob, we already
+        // have their documents and should truncate the vector of readers.
+        if first_active > 0 && first_active < readers.len() {
+            readers.range(first_active..);
+        }
```

Comment:
I don't think we need the check. Also doesn't seem we're making the use the result of the range here.